omp/solver/gmres_kernels.cpp

```diff
@@ -48,6 +49,200 @@ namespace omp {
 namespace gmres {
 
 
+namespace {
+
+
+template <typename ValueType>
+void finish_arnoldi(matrix::Dense<ValueType> *next_krylov_basis,
+                    matrix::Dense<ValueType> *krylov_bases,
+                    matrix::Dense<ValueType> *hessenberg_iter, size_type iter,
+                    const stopping_status *stop_status)
+{
+#pragma omp declare reduction(add : ValueType : omp_out = omp_out + omp_in)
+
+    for (size_type i = 0; i < next_krylov_basis->get_size()[1]; ++i) {
+        if (stop_status[i].has_stopped()) {
+            continue;
+        }
+        for (size_type k = 0; k < iter + 1; ++k) {
+            ValueType hessenberg_iter_entry = 0;
+
+#pragma omp parallel for reduction(add : hessenberg_iter_entry)
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                hessenberg_iter_entry +=
+                    next_krylov_basis->at(j, i) *
+                    krylov_bases->at(j,
+                                     next_krylov_basis->get_size()[1] * k + i);
+            }
+            hessenberg_iter->at(k, i) = hessenberg_iter_entry;
+
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                next_krylov_basis->at(j, i) -=
+                    hessenberg_iter->at(k, i) *
+                    krylov_bases->at(j,
+                                     next_krylov_basis->get_size()[1] * k + i);
+            }
+        }
+        // for i in 1:iter
+        //     hessenberg(iter, i) = next_krylov_basis' * krylov_bases(:, i)
+        //     next_krylov_basis  -= hessenberg(iter, i) * krylov_bases(:, i)
+        // end
+
+        ValueType hessenberg_iter_entry = 0;
+
+#pragma omp parallel for reduction(add : hessenberg_iter_entry)
+        for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
```

Comment:
Same here: the krylov subspace size is usually small (<100) and parallelizing requires a global reduction...